homeassistant/components/zwave_js/config_flow.py

```diff
@@ -62,14 +110,268 @@ async def async_step_user(
             _LOGGER.exception("Unexpected exception")
             errors["base"] = "unknown"
         else:
-            await self.async_set_unique_id(version_info.home_id)
+            await self.async_set_unique_id(
+                version_info.home_id, raise_on_progress=False
+            )
             self._abort_if_unique_id_configured(user_input)
-            return self.async_create_entry(title=NAME, data=user_input)
+            self.ws_address = user_input[CONF_URL]
+            return self._async_create_entry_from_vars()
 
         return self.async_show_form(
-            step_id="user", data_schema=STEP_USER_DATA_SCHEMA, errors=errors
+            step_id="manual", data_schema=STEP_USER_DATA_SCHEMA, errors=errors
+        )
+
+    async def async_step_hassio(  # type: ignore
+        self, discovery_info: Dict[str, Any]
+    ) -> Dict[str, Any]:
+        """Receive configuration from add-on discovery info.
+
+        This flow is triggered by the Z-Wave JS add-on.
+        """
+        assert self.hass
+        self.ws_address = f"ws://{discovery_info['host']}:{discovery_info['port']}"
+        try:
+            version_info = await async_get_version_info(self.hass, self.ws_address)
+        except CannotConnect:
+            return self.async_abort(reason="cannot_connect")
+
+        await self.async_set_unique_id(version_info.home_id)
+        self._abort_if_unique_id_configured(updates={CONF_URL: self.ws_address})
+
+        return await self.async_step_hassio_confirm()
+
+    async def async_step_hassio_confirm(
+        self, user_input: Optional[Dict[str, Any]] = None
+    ) -> Dict[str, Any]:
+        """Confirm the add-on discovery."""
+        if user_input is not None:
+            return await self.async_step_on_supervisor(
+                user_input={CONF_USE_ADDON: True}
+            )
+
+        return self.async_show_form(step_id="hassio_confirm")
+
+    def _async_create_entry_from_vars(self) -> Dict[str, Any]:
+        """Return a config entry for the flow."""
+        return self.async_create_entry(
+            title=TITLE,
+            data={
+                CONF_URL: self.ws_address,
+                CONF_USB_PATH: self.usb_path,
+                CONF_NETWORK_KEY: self.network_key,
```

Comment:
Why would we store that?

Reply:
To put the integration in control? If the add-on is uninstalled or modified in whatever way, the integration can fix/restore it. (at least that was the original scope of the discussion we had for OZW about this).

Reply:
The old plan was to be able to update add-on configuration from the integration at a later stage, I believe.

Reply:
For updating the config, it doesn't matter because we can just read the config out when we need to show it to the user. 

But making the integration the source of truth == 👍 . Would be great if we could mark add-ons as `system-managed` and so can only be removed via config entry. But that's for future.